File: T2/EB_BDD_GR9/03.RESTFULL_JAVA_EUREKABANK_GR09/03.RESTFULL_JAVA_EUREKABANK_GR09/02.CLIESC/BDD_RESTFUL_JAVA_CLIESC_G09/src/main/java/ec/edu/monster/vista/TransferenciaView.java
/*
 * Vista para realizar transferencias - Diseño mejorado
 * @author Grupo Monster G09
 */
package ec.edu.monster.vista;

import ec.edu.monster.controlador.Desktop_Controlador;
import javafx.geometry.Insets;
import javafx.geometry.Pos;
import javafx.scene.Scene;
import javafx.scene.control.Alert;
import javafx.scene.control.Button;
import javafx.scene.control.Label;
import javafx.scene.control.TextField;
import javafx.scene.layout.BorderPane;
import javafx.scene.layout.HBox;
import javafx.scene.layout.StackPane;
import javafx.scene.layout.VBox;
import javafx.stage.Stage;

/**
 * Vista para realizar transferencias entre cuentas con diseño mejorado
 */
public class TransferenciaView {
    
    private static final String COLOR_PRIMARY = "#3ab4d9";
    private static final String COLOR_YELLOW = "#f6de88";
    private static final String COLOR_BACKGROUND_LIGHT = "#afe0f8";
    private static final String COLOR_TEXT_DARK = "#0f171a";
    
    private Desktop_Controlador controlador;
    
    public void mostrar(Stage stage, Desktop_Controlador controlador) {
        this.controlador = controlador;
        
        stage.setTitle("EurekaBank - Realizar Transferencia");
        stage.setWidth(1000);
        stage.setHeight(800);
        
        BorderPane root = new BorderPane();
        root.setStyle("-fx-background-color: " + COLOR_BACKGROUND_LIGHT + ";");
        
        // Header
        HBox header = crearHeader(stage);
        root.setTop(header);
        
        // Contenido
        VBox content = crearContenido(stage);
        root.setCenter(content);
        
        Scene scene = new Scene(root);
        stage.setScene(scene);
    }
    
    /**
     * Crea el header mejorado
     */
    private HBox crearHeader(Stage stage) {
        HBox header = new HBox(20);
        header.setPadding(new Insets(12, 40, 12, 40));
        header.setStyle(
            "-fx-background-color: rgba(255,255,255,0.5); " +
            "-fx-border-color: rgba(255,255,255,0.3); " +
            "-fx-border-width: 0 0 1 0;"
        );
        header.setAlignment(Pos.CENTER_LEFT);
        
        Label logo = new Label("EurekaBank");
        logo.setStyle("-fx-font-size: 24px; -fx-font-weight: bold; -fx-text-fill: " + COLOR_TEXT_DARK + ";");
        
        HBox spacer = new HBox();
        javafx.scene.layout.Region region = new javafx.scene.layout.Region();
        HBox.setHgrow(region, javafx.scene.layout.Priority.ALWAYS);
        spacer.getChildren().add(region);
        
        Button btnVolver = new Button("Volver");
        btnVolver.setStyle(
            "-fx-font-size: 14px; " +
            "-fx-pref-width: 80px; " +
            "-fx-pref-height: 36px; " +
            "-fx-background-color: " + COLOR_PRIMARY + "; " +
            "-fx-text-fill: white; " +
            "-fx-background-radius: 6px; " +
            "-fx-cursor: hand;"
        );
        btnVolver.setOnAction(e -> {
            MenuView menuView = new MenuView();
            menuView.mostrar(stage);
        });
        
        header.getChildren().addAll(logo, spacer, btnVolver);
        
        return header;
    }
    
    /**
     * Crea el contenido del formulario mejorado
     */
    private VBox crearContenido(Stage stage) {
        VBox content = new VBox(15);
        content.setPadding(new Insets(20, 40, 20, 40));
        content.setAlignment(Pos.CENTER);
        content.setMaxWidth(700);
        
        // Tarjeta principal
        VBox card = new VBox(20);
        card.setPadding(new Insets(30, 40, 30, 40));
        card.setStyle(
            "-fx-background-color: white; " +
            "-fx-background-radius: 16px; " +
            "-fx-effect: dropshadow(gaussian, rgba(0,0,0,0.1), 12, 0, 0, 4);"
        );
        card.setAlignment(Pos.CENTER);
        
        // Icono decorativo (más pequeño)
        javafx.scene.shape.Circle iconCircle = new javafx.scene.shape.Circle(35);
        iconCircle.setFill(javafx.scene.paint.Color.web(COLOR_YELLOW, 0.2));
        iconCircle.setStroke(javafx.scene.paint.Color.web(COLOR_YELLOW));
        iconCircle.setStrokeWidth(3);
        
        StackPane iconPane = new StackPane(iconCircle);
        iconPane.setAlignment(Pos.CENTER);
        iconPane.setPadding(new Insets(0, 0, 5, 0));
        
        // Título (más pequeño)
        Label title = new Label("Realizar Transferencia");
        title.setStyle("-fx-font-size: 28px; -fx-font-weight: 900; -fx-text-fill: " + COLOR_TEXT_DARK + ";");
        
        Label subtitle = new Label("Envía dinero a otras cuentas de forma segura");
        subtitle.setStyle("-fx-font-size: 14px; -fx-text-fill: #6b7280;");
        
        // Formulario
        VBox form = new VBox(16);
        form.setPrefWidth(600);
        
        // Campo Cuenta Origen
        VBox cuentaOrigenBox = new VBox(8);
        Label cuentaOrigenLabel = new Label("Cuenta Origen");
        cuentaOrigenLabel.setStyle("-fx-font-size: 16px; -fx-font-weight: 500; -fx-text-fill: " + COLOR_TEXT_DARK + ";");
        
        TextField txtCuentaOrigen = new TextField();
        txtCuentaOrigen.setPromptText("Ingrese el número de cuenta origen");
        txtCuentaOrigen.setStyle(
            "-fx-font-size: 15px; " +
            "-fx-pref-height: 46px; " +
            "-fx-padding: 10px 16px; " +
            "-fx-background-color: #f6f7f8; " +
            "-fx-border-color: #d1d5db; " +
            "-fx-border-radius: 8px; " +
            "-fx-background-radius: 8px; " +
            "-fx-effect: dropshadow(gaussian, rgba(0,0,0,0.05), 4, 0, 0, 2);"
        );
        
        cuentaOrigenBox.getChildren().addAll(cuentaOrigenLabel, txtCuentaOrigen);
        
        // Campo Cuenta Destino
        VBox cuentaDestinoBox = new VBox(8);
        Label cuentaDestinoLabel = new Label("Cuenta Destino");
        cuentaDestinoLabel.setStyle("-fx-font-size: 16px; -fx-font-weight: 500; -fx-text-fill: " + COLOR_TEXT_DARK + ";");
        
        TextField txtCuentaDestino = new TextField();
        txtCuentaDestino.setPromptText("Ingrese el número de cuenta destino");
        txtCuentaDestino.setStyle(
            "-fx-font-size: 15px; " +
            "-fx-pref-height: 46px; " +
            "-fx-padding: 10px 16px; " +
            "-fx-background-color: #f6f7f8; " +
            "-fx-border-color: #d1d5db; " +
            "-fx-border-radius: 8px; " +
            "-fx-background-radius: 8px; " +
            "-fx-effect: dropshadow(gaussian, rgba(0,0,0,0.05), 4, 0, 0, 2);"
        );
        
        cuentaDestinoBox.getChildren().addAll(cuentaDestinoLabel, txtCuentaDestino);
        
        // Campo Importe
        VBox importeBox = new VBox(8);
        Label importeLabel = new Label("Importe");
        importeLabel.setStyle("-fx-font-size: 16px; -fx-font-weight: 500; -fx-text-fill: " + COLOR_TEXT_DARK + ";");
        
        TextField txtImporte = new TextField();
        txtImporte.setPromptText("Ingrese el importe a transferir");
        txtImporte.setStyle(
            "-fx-font-size: 15px; " +
            "-fx-pref-height: 46px; " +
            "-fx-padding: 10px 16px; " +
            "-fx-background-color: #f6f7f8; " +
            "-fx-border-color: #d1d5db; " +
            "-fx-border-radius: 8px; " +
            "-fx-background-radius: 8px; " +
            "-fx-effect: dropshadow(gaussian, rgba(0,0,0,0.05), 4, 0, 0, 2);"
        );
        
        importeBox.getChildren().addAll(importeLabel, txtImporte);
        
        // Botón de transferencia (más pequeño)
        Button btnTransferir = new Button("Realizar Transferencia");
        btnTransferir.setStyle(
            "-fx-font-size: 16px; " +
            "-fx-font-weight: bold; " +
            "-fx-pref-width: 600px; " +
            "-fx-pref-height: 48px; " +
            "-fx-background-color: " + COLOR_YELLOW + "; " +
            "-fx-text-fill: " + COLOR_TEXT_DARK + "; " +
            "-fx-background-radius: 8px; " +
            "-fx-cursor: hand; " +
            "-fx-effect: dropshadow(gaussian, rgba(246,222,136,0.3), 8, 0, 0, 4);"
        );
        
        btnTransferir.setOnMouseEntered(e -> btnTransferir.setStyle(
            "-fx-font-size: 16px; " +
            "-fx-font-weight: bold; " +
            "-fx-pref-width: 600px; " +
            "-fx-pref-height: 48px; " +
            "-fx-background-color: #e5c868; " +
            "-fx-text-fill: " + COLOR_TEXT_DARK + "; " +
            "-fx-background-radius: 8px; " +
            "-fx-cursor: hand; " +
            "-fx-effect: dropshadow(gaussian, rgba(246,222,136,0.3), 8, 0, 0, 4);"
        ));
        btnTransferir.setOnMouseExited(e -> btnTransferir.setStyle(
            "-fx-font-size: 16px; " +
            "-fx-font-weight: bold; " +
            "-fx-pref-width: 600px; " +
            "-fx-pref-height: 48px; " +
            "-fx-background-color: " + COLOR_YELLOW + "; " +
            "-fx-text-fill: " + COLOR_TEXT_DARK + "; " +
            "-fx-background-radius: 8px; " +
            "-fx-cursor: hand; " +
            "-fx-effect: dropshadow(gaussian, rgba(246,222,136,0.3), 8, 0, 0, 4);"
        ));
        
        btnTransferir.setOnAction(e -> {
            String cuentaOrigen = txtCuentaOrigen.getText().trim();
            String cuentaDestino = txtCuentaDestino.getText().trim();
            String importeStr = txtImporte.getText().trim();
            
            if (cuentaOrigen.isEmpty() || cuentaDestino.isEmpty() || importeStr.isEmpty()) {
                mostrarError("Por favor, complete todos los campos.");
                return;
            }
            
            if (cuentaOrigen.equals(cuentaDestino)) {
                mostrarError("No se puede transferir a la misma cuenta.");
                return;
            }
            
            try {
                double importe = Double.parseDouble(importeStr.replace(",", "."));
                if (importe <= 0) {
                    mostrarError("El importe debe ser mayor a cero.");
                    return;
                }
                
                int resultado = controlador.regTransferencia(cuentaOrigen, cuentaDestino, importe);
                if (resultado == 1) {
                    Alert alert = new Alert(Alert.AlertType.INFORMATION);
                    alert.setTitle("Éxito");
                    alert.setHeaderText(null);
                    alert.setContentText("Transferencia realizada exitosamente.");
                    alert.showAndWait();
                    
                    // Limpiar campos
                    txtCuentaOrigen.clear();
                    txtCuentaDestino.clear();
                    txtImporte.clear();
                } else {
                    mostrarError("Error al procesar la transferencia. Verifique los datos.");
                }
            } catch (NumberFormatException ex) {
                mostrarError("Formato inválido. Ingrese un número válido.");
            }
        });
        
        form.getChildren().addAll(cuentaOrigenBox, cuentaDestinoBox, importeBox, btnTransferir);
        
        card.getChildren().addAll(iconPane, title, subtitle, form);
        
        content.getChildren().addAll(card);
        
        return content;
    }
    
    /**
     * Muestra un mensaje de error
     */
    private void mostrarError(String mensaje) {
        Alert alert = new Alert(Alert.AlertType.ERROR);
        alert.setTitle("Error");
        alert.setHeaderText(null);
        alert.setContentText(mensaje);
        alert.showAndWait();
    }
}

File: T2/EB_BDD_GR9/04._RESTFUL_DOTNET_EUREKABANK_GR09/04._RESTFUL_DOTNET_EUREKABANK_GR09/02. CLIESC/BDD_RESTFUL_DOTNET_CLIESC_G09/BDD_RESTFUL_DOTNET_CLIESC_G09/ec/edu/monster/vista/TransferenciaView.cs
using System;
using System.Drawing;
using System.Drawing.Drawing2D;
using System.Windows.Forms;
using BDD_RESTFUL_DOTNET_CLIESC_G09.ec.edu.monster.controlador;

namespace BDD_RESTFUL_DOTNET_CLIESC_G09.ec.edu.monster.vista
{
    /// <summary>
    /// Vista para realizar transferencias - Diseño mejorado (igual al Java)
    /// </summary>
    public partial class TransferenciaView : Form
    {
        private static readonly Color COLOR_PRIMARY = Color.FromArgb(58, 180, 217);      // #3ab4d9
        private static readonly Color COLOR_YELLOW = Color.FromArgb(246, 222, 136);       // #f6de88
        private static readonly Color COLOR_BACKGROUND_LIGHT = Color.FromArgb(175, 224, 248); // #afe0f8
        private static readonly Color COLOR_TEXT_DARK = Color.FromArgb(15, 23, 26);      // #0f171a

        private Desktop_Controlador controlador;
        private TextBox txtCuentaOrigen;
        private TextBox txtCuentaDestino;
        private TextBox txtImporte;

        public TransferenciaView()
        {
            InitializeComponent();
            controlador = new Desktop_Controlador();
        }

        private void InitializeComponent()
        {
            this.SuspendLayout();

            this.Text = "EurekaBank - Realizar Transferencia";
            this.Size = new Size(1000, 800);
            this.StartPosition = FormStartPosition.CenterScreen;
            this.BackColor = COLOR_BACKGROUND_LIGHT;
            this.FormBorderStyle = FormBorderStyle.FixedSingle;
            this.MaximizeBox = false;

            Panel header = CrearHeader();
            this.Controls.Add(header);

            Panel content = new Panel();
            content.Size = new Size(1000, 740);
            content.Location = new Point(0, 60);
            content.BackColor = COLOR_BACKGROUND_LIGHT;
            content.Padding = new Padding(40, 20, 40, 20);

            Panel card = new Panel();
            card.Size = new Size(700, 600);
            card.Location = new Point(150, 50);
            card.BackColor = Color.White;
            card.Padding = new Padding(40, 30, 40, 30);
            card.Paint += Card_Paint;

            Label title = new Label();
            title.Text = "Realizar Transferencia";
            title.Font = new Font("Segoe UI", 28, FontStyle.Bold);
            title.ForeColor = COLOR_TEXT_DARK;
            title.AutoSize = true;
            title.Location = new Point(200, 20);

            Label subtitle = new Label();
            subtitle.Text = "Envía dinero a otras cuentas de forma segura";
            subtitle.Font = new Font("Segoe UI", 14);
            subtitle.ForeColor = Color.FromArgb(107, 114, 128);
            subtitle.AutoSize = true;
            subtitle.Location = new Point(150, 60);

            // Campo Cuenta Origen
            Label cuentaOrigenLabel = new Label();
            cuentaOrigenLabel.Text = "Cuenta Origen";
            cuentaOrigenLabel.Font = new Font("Segoe UI", 16, FontStyle.Regular);
            cuentaOrigenLabel.ForeColor = COLOR_TEXT_DARK;
            cuentaOrigenLabel.AutoSize = true;
            cuentaOrigenLabel.Location = new Point(40, 140);

            txtCuentaOrigen = new TextBox();
            txtCuentaOrigen.Font = new Font("Segoe UI", 15);
            txtCuentaOrigen.Size = new Size(600, 46);
            txtCuentaOrigen.Location = new Point(40, 170);
            txtCuentaOrigen.BackColor = Color.FromArgb(246, 247, 248);
            txtCuentaOrigen.BorderStyle = BorderStyle.FixedSingle;

            // Campo Cuenta Destino
            Label cuentaDestinoLabel = new Label();
            cuentaDestinoLabel.Text = "Cuenta Destino";
            cuentaDestinoLabel.Font = new Font("Segoe UI", 16, FontStyle.Regular);
            cuentaDestinoLabel.ForeColor = COLOR_TEXT_DARK;
            cuentaDestinoLabel.AutoSize = true;
            cuentaDestinoLabel.Location = new Point(40, 240);

            txtCuentaDestino = new TextBox();
            txtCuentaDestino.Font = new Font("Segoe UI", 15);
            txtCuentaDestino.Size = new Size(600, 46);
            txtCuentaDestino.Location = new Point(40, 270);
            txtCuentaDestino.BackColor = Color.FromArgb(246, 247, 248);
            txtCuentaDestino.BorderStyle = BorderStyle.FixedSingle;

            // Campo Importe
            Label importeLabel = new Label();
            importeLabel.Text = "Importe";
            importeLabel.Font = new Font("Segoe UI", 16, FontStyle.Regular);
            importeLabel.ForeColor = COLOR_TEXT_DARK;
            importeLabel.AutoSize = true;
            importeLabel.Location = new Point(40, 340);

            txtImporte = new TextBox();
            txtImporte.Font = new Font("Segoe UI", 15);
            txtImporte.Size = new Size(600, 46);
            txtImporte.Location = new Point(40, 370);
            txtImporte.BackColor = Color.FromArgb(246, 247, 248);
            txtImporte.BorderStyle = BorderStyle.FixedSingle;

            // Botón de transferencia (600px ancho, yellow, texto dark)
            Button btnTransferir = new Button();
            btnTransferir.Text = "Realizar Transferencia";
            btnTransferir.Font = new Font("Segoe UI", 16, FontStyle.Bold);
            btnTransferir.Size = new Size(600, 48);
            btnTransferir.Location = new Point(40, 450);
            btnTransferir.BackColor = COLOR_YELLOW;
            btnTransferir.ForeColor = COLOR_TEXT_DARK;
            btnTransferir.FlatStyle = FlatStyle.Flat;
            btnTransferir.FlatAppearance.BorderSize = 0;
            btnTransferir.Cursor = Cursors.Hand;
            btnTransferir.Click += BtnTransferir_Click;
            btnTransferir.MouseEnter += (s, e) => btnTransferir.BackColor = Color.FromArgb(229, 200, 104); // #e5c868
            btnTransferir.MouseLeave += (s, e) => btnTransferir.BackColor = COLOR_YELLOW;

            card.Controls.Add(title);
            card.Controls.Add(subtitle);
            card.Controls.Add(cuentaOrigenLabel);
            card.Controls.Add(txtCuentaOrigen);
            card.Controls.Add(cuentaDestinoLabel);
            card.Controls.Add(txtCuentaDestino);
            card.Controls.Add(importeLabel);
            card.Controls.Add(txtImporte);
            card.Controls.Add(btnTransferir);

            content.Controls.Add(card);
            this.Controls.Add(content);

            this.ResumeLayout(false);
        }

        private void Card_Paint(object sender, PaintEventArgs e)
        {
            Graphics g = e.Graphics;
            g.SmoothingMode = SmoothingMode.AntiAlias;

            // Icono decorativo: Circle 35, fill yellow 0.2, stroke yellow, strokeWidth 3
            using (Pen pen = new Pen(COLOR_YELLOW, 3))
            using (SolidBrush brush = new SolidBrush(Color.FromArgb(51, COLOR_YELLOW))) // 0.2 opacity
            {
                g.FillEllipse(brush, 315, 20, 70, 70);
                g.DrawEllipse(pen, 315, 20, 70, 70);
            }
        }

        private Panel CrearHeader()
        {
            Panel header = new Panel();
            header.Size = new Size(1000, 60);
            header.Location = new Point(0, 0);
            header.BackColor = Color.FromArgb(128, 255, 255, 255);
            header.Paint += (s, e) => {
                Graphics g = e.Graphics;
                using (Pen pen = new Pen(Color.FromArgb(77, 255, 255, 255), 1))
                {
                    g.DrawLine(pen, 0, 59, 1000, 59);
                }
                using (Font font = new Font("Segoe UI", 24, FontStyle.Bold))
                using (SolidBrush brush = new SolidBrush(COLOR_TEXT_DARK))
                {
                    g.DrawString("EurekaBank", font, brush, 40, 15);
                }
            };

            Button btnVolver = new Button();
            btnVolver.Text = "Volver";
            btnVolver.Font = new Font("Segoe UI", 14);
            btnVolver.Size = new Size(80, 36);
            btnVolver.Location = new Point(900, 12);
            btnVolver.BackColor = COLOR_PRIMARY;
            btnVolver.ForeColor = Color.White;
            btnVolver.FlatStyle = FlatStyle.Flat;
            btnVolver.FlatAppearance.BorderSize = 0;
            btnVolver.Cursor = Cursors.Hand;
            btnVolver.Click += (s, e) => {
                MenuView menuView = new MenuView();
                menuView.Show();
                this.Close();
            };
            header.Controls.Add(btnVolver);

            return header;
        }

        private void BtnTransferir_Click(object sender, EventArgs e)
        {
            string cuentaOrigen = txtCuentaOrigen.Text.Trim();
            string cuentaDestino = txtCuentaDestino.Text.Trim();
            string importeStr = txtImporte.Text.Trim();

            if (string.IsNullOrEmpty(cuentaOrigen) || string.IsNullOrEmpty(cuentaDestino) || string.IsNullOrEmpty(importeStr))
            {
                MessageBox.Show("Por favor, complete todos los campos.", "Error", 
                    MessageBoxButtons.OK, MessageBoxIcon.Error);
                return;
            }

            if (cuentaOrigen.Equals(cuentaDestino))
            {
                MessageBox.Show("No se puede transferir a la misma cuenta.", "Error", 
                    MessageBoxButtons.OK, MessageBoxIcon.Error);
                return;
            }

            try
            {
                double importe = double.Parse(importeStr.Replace(",", "."));
                if (importe <= 0)
                {
                    MessageBox.Show("El importe debe ser mayor a cero.", "Error", 
                        MessageBoxButtons.OK, MessageBoxIcon.Error);
                    return;
                }

                int resultado = controlador.RegTransferencia(cuentaOrigen, cuentaDestino, importe);
                if (resultado == 1)
                {
                    MessageBox.Show("Transferencia realizada exitosamente.", "Éxito", 
                        MessageBoxButtons.OK, MessageBoxIcon.Information);
                    txtCuentaOrigen.Clear();
                    txtCuentaDestino.Clear();
                    txtImporte.Clear();
                }
                else
                {
                    MessageBox.Show("Error al procesar la transferencia. Verifique los datos.", "Error", 
                        MessageBoxButtons.OK, MessageBoxIcon.Error);
                }
            }
            catch (FormatException)
            {
                MessageBox.Show("Formato inválido. Ingrese un número válido.", "Error", 
                    MessageBoxButtons.OK, MessageBoxIcon.Error);
            }
        }
    }
}



